src/apps/rack_level_slots/services.py

```diff
@@ -0,0 +1,400 @@
+    output_schema: BaseModel = RackLevelSlotBaseOutputSchema,
+) -> Union[
+    PagedResponseSchema[RackLevelSlotBaseOutputSchema],
+    PagedResponseSchema[RackLevelSlotOutputSchema],
+]:
+    query = select(RackLevelSlot)
+
+    return await paginate(
+        query=query,
+        response_schema=output_schema,
+        table=RackLevelSlot,
+        page_params=page_params,
+        session=session,
+    )
+
+
+async def update_single_rack_level_slot(
+    session: AsyncSession,
+    rack_level_slot_input: RackLevelSlotUpdateSchema,
+    rack_level_slot_id: str,
+) -> RackLevelSlotOutputSchema:
+    if not (
+        rack_level_slot_object := await if_exists(
+            RackLevelSlot, "id", rack_level_slot_id, session
+        )
+    ):
+        raise DoesNotExist(RackLevelSlot.__name__, "id", rack_level_slot_id)
+
+    rack_level_slot_data = rack_level_slot_input.dict(
+        exclude_unset=True, exclude_none=True
+    )
+
+    if rack_level_slot_data:
+        statement = (
+            update(RackLevelSlot)
+            .filter(RackLevelSlot.id == rack_level_slot_id)
+            .values(**rack_level_slot_data)
+        )
+
+        await session.execute(statement)
+        await session.commit()
+        await session.refresh(rack_level_slot_object)
+
+    return await get_single_rack_level_slot(
+        session, rack_level_slot_id=rack_level_slot_id
+    )
+
+
+async def manage_single_rack_level_slot_state(
+    session: AsyncSession, rack_level_slot_id: str, activate_slot: bool = True
+) -> dict[str, str]:
+    from src.apps.rack_levels.services import manage_rack_level_state
+
+    if not (
+        rack_level_slot_object := await if_exists(
+            RackLevelSlot, "id", rack_level_slot_id, session
+        )
+    ):
+        raise DoesNotExist(RackLevelSlot.__name__, "id", rack_level_slot_id)
+
+    if rack_level_slot_object.stock:
+        raise RackLevelSlotIsNotEmptyException(resource="Slot contains a stock! ")
+
+    if rack_level_slot_object.is_active and activate_slot:
+        raise CantActivateRackLevelSlotException(reason="Slot already activated! ")
+
+    if not (rack_level_slot_object.is_active) and not (activate_slot):
+        raise CantDeactivateRackLevelSlotException(reason="Slot already deactivated! ")
+
+    if not (
+        rack_level_object := await if_exists(
+            RackLevel, "id", rack_level_slot_object.rack_level_id, session
+        )
+    ):
+        raise DoesNotExist(
+            RackLevel.__name__, "id", rack_level_slot_object.rack_level_id
+        )
+
+    if (not rack_level_object.available_slots) and (not activate_slot):
+        raise CantDeactivateRackLevelSlotException(
+            reason="All available slots are occupied! "
+        )
+
+    if (
+        rack_level_object.active_slots == rack_level_object.max_slots
+    ) and activate_slot:
+        raise CantActivateRackLevelSlotException(
+            reason="Reached max amount of activated slots in the rack level! "
+        )
```

Comment:
it seems to be useless - there is no case when it will appear - active slots = available + occupied
occupied is not empty and active and wont be activated again, available is empty but active - wont be activated again